### PR TITLE
Document /dss/files and /fetch/dss/files endpoints (#1402)

### PR DESCRIPTION
**Pull request by noah-aviel-dove**
The /fetch endpoint didn't have its own ticket so I figured I'd cover it here.

### CODE REVIEW

File: lambdas/service/app.py [hunk 1]
```diff
@@ -1062,7 +1143,7 @@ def fetch_dss_files(uuid):
 
 
 def _dss_files(uuid, fetch=True):
-    query_params = app.current_request.query_params
+    query_params = app.current_request.query_params or {}
```
Comment (achave11) [inline lambdas/service/app.py]: This change isn't specific to the documentation, should it be in a different commit? 
Also this seems like a regression from the present implementation, I could be wrong.

File: lambdas/service/app.py [hunk 2]
```diff
@@ -85,6 +85,10 @@
         {
             'name': 'Health',
             'description': 'For checking on service health'
+        },
+        {
+            'name': 'DSS',
```
Comment (achave11) [inline lambdas/service/app.py]: We might want to be explicit about this endpoints being a `get` file proxy from the DSS aws replica.

File: lambdas/service/app.py [hunk 3]
```diff
@@ -925,7 +929,80 @@ def generate_manifest(event, context):
     return {'Location': presigned_url}
 
 
-@app.route('/dss/files/{uuid}', methods=['GET'], cors=True)
+dss_files_spec = {
+    'description': 'Request a signed download link to a file from the Data Store',
```
Comment (achave11) [inline lambdas/service/app.py]: We should make the distinction of the `/fetch/dss/files/{uuid}` and `/dss/files/{uuid} ` endpoint descriptions since it is a significant one.

File: lambdas/service/app.py (hunk 1 above)
Comment (noah-aviel-dove) [reply lambdas/service/app.py]: It IS in its own commit. Did you mean a different pull request?

Comment (achave11) [reply lambdas/service/app.py]: Yea that was a mistake.

Comment (achave11) [reply lambdas/service/app.py]: Yea that was a mistake on my behalf.

File: lambdas/service/app.py [hunk 4]
```diff
@@ -925,7 +929,80 @@ def generate_manifest(event, context):
     return {'Location': presigned_url}
 
 
-@app.route('/dss/files/{uuid}', methods=['GET'], cors=True)
+dss_files_spec = {
```
Comment (achave11) [inline lambdas/service/app.py]: You can add the 'tags' field in this dict to reduce duplication.

**Comment by achave11** [comment]
For the next review cycle, ensure the `Try it out` example is functional. Also please fill free to talk to me about any comments or questions you might have.

**Comment by noah-aviel-dove** [comment]
There appears to be an issue with the "Try it out" functionality with the `/dss/files/` endpoint. Swagger follows redirects, meaning that instead of seeing the 30X responses, it spins for a seconds before redirecting all the way to the requested file and displaying its contents with a 200 response. 

While this is an accurate demonstration of how the endpoint behaves, it appears to conflict with the documented responses, since we never see the 30X. I'm not sure whether this should be added to the documentation or not.

File: src/azul/openapi/schema.py [hunk 5]
```diff
@@ -117,6 +122,90 @@ def array(item: TYPE, *items: TYPE, **kwargs):
     return array_type(make_type(item), *map(make_type, items), **kwargs)
 
 
+def enum_type(items: Iterable) -> JSON:
```
Comment (hannes-ucsc) [inline src/azul/openapi/schema.py]: This should use varargs like other helpers. The name should be just `enum`.

@achave11 is working on this too. You guys should coordinate.

I like the implicit type determination but there should be a way to explicitly pass the type.

File: src/azul/openapi/schema.py [hunk 6]
```diff
@@ -117,6 +122,90 @@ def array(item: TYPE, *items: TYPE, **kwargs):
...
+    return make_param(name, 'query', type_, required=not is_optional, description=description)
+
+
+def header_response(type_: TYPE, description: Optional[str] = None) -> JSON:
+    """
+    Returns the schema and description for a response header.
+    """
+    return {
+        'schema': make_type(type_),
+        **({} if description is None else {'description': description})
```
Comment (hannes-ucsc) [inline src/azul/openapi/schema.py]: This conditional can be eliminated simply by allowing and interpolating `**kwargs` like some other helpers do. 

This is not supposed to be an abstraction over OpenAPI spec, but merely a way to reduce its verbosity when specified as a JSON literal.

File: src/azul/openapi/schema.py [hunk 7]
```diff
@@ -220,3 +309,27 @@ def make_type(t: TYPE) -> JSON:
         return t
     else:
         assert False, type(t)
+
+
+def make_param(name: str, in_: str, type_: TYPE, required: bool, description: Optional[str]) -> JSON:
```
Comment (hannes-ucsc) [inline src/azul/openapi/schema.py]:
```suggestion
def make_param(name: str, in_: str, type_: TYPE, optional: bool, description: Optional[str]) -> JSON:
```

File: src/azul/openapi/schema.py [hunk 8]
```diff
@@ -117,6 +122,90 @@ def array(item: TYPE, *items: TYPE, **kwargs):
...
+            "foo"
+        ]
+    }
+    """
+    items = sorted(items)
+    type_ = one(set(map(type, items)))
+    return dict(make_type(type_), enum=items)
+
+
+def path_param(name: str, type_: TYPE, description: Optional[str] = None) -> JSON:
```
Comment (hannes-ucsc) [inline src/azul/openapi/schema.py]: The param stuff should be in a separate `params` module. `schema` is for the JSONSchema subset of OpenAPI.

Instead of `schema.path_param(…)` an invocation would be `params.path` which is a bit shorter.

File: src/azul/openapi/schema.py [hunk 9]
```diff
@@ -220,3 +309,27 @@ def make_type(t: TYPE) -> JSON:
...
+        },
+        "description": "razzle that dazzle"
+    }
+    """
+    return {
+        'name': name,
+        'in': in_,
+        'required': required,
+        'schema': make_type(type_),
+        **({} if description is None else {'description': description})
```
Comment (hannes-ucsc) [inline src/azul/openapi/schema.py]: **kwargs

File: src/azul/openapi/schema.py [hunk 10]
```diff
@@ -220,3 +309,27 @@ def make_type(t: TYPE) -> JSON:
         return t
     else:
         assert False, type(t)
+
+
+def make_param(name: str, in_: str, type_: TYPE, required: bool, description: Optional[str]) -> JSON:
+    """
+    >>> from azul.doctests import assert_json
+    >>> assert_json(make_param('foo', 'query', int, False, 'razzle that dazzle'))
```
Comment (hannes-ucsc) [inline src/azul/openapi/schema.py]: Because of the large number of heterogeneous positional args, this invocation is hard to read if it were to occur in an actual spec. To make it easier to read, one would need to pass most of the positionals as keywords anyway which wouldn't be much shorter than the raw JSON literal. OTOH, having this helper enables the IDE's type checker to warn about missing arguments and type mismatches. I think we should leave this helper in, but make it protected and convert all calls to use keyword argument passing even for positional arguments.

File: src/azul/openapi/schema.py [hunk 11]
```diff
@@ -117,6 +122,90 @@ def array(item: TYPE, *items: TYPE, **kwargs):
...
+        "required": false,
+        "schema": {
+            "type": "integer",
+            "format": "int64"
+        }
+    }
+    """
+    is_optional = isinstance(type_, optional)
+    if is_optional:
+        type_ = type_.property
```
Comment (hannes-ucsc) [inline src/azul/openapi/schema.py]: I think we should rename `optional.property` to `optional.type_`

**Comment by achave11** [comment]
@noah-aviel-dove As described [here](https://github.com/swagger-api/swagger-ui/issues/1569#issuecomment-390877095), swagger-ui follows redirects.

File: src/azul/openapi/schema.py [hunk 12]
```diff
@@ -117,6 +122,29 @@ def array(item: TYPE, *items: TYPE, **kwargs):
...
+    >>> assert_json(enum('foo', 'bar', type_=str))
+    {
+        "type": "string",
+        "enum": [
+            "foo",
+            "bar"
+        ]
+    }
+    """
+    if type_:
```
Comment (achave11) [inline src/azul/openapi/schema.py]: I will need to address this [requested changes](https://github.com/DataBiosphere/azul/pull/1538/files#r385474794). I am at fault for this.

File: src/azul/openapi/schema.py [hunk 13]
```diff
@@ -117,6 +122,51 @@ def array(item: TYPE, *items: TYPE, **kwargs):
     return array_type(make_type(item), *map(make_type, items), **kwargs)
 
 
+def enum(*items: PrimitiveJSON, type_: TYPE = None) -> JSON:
+    """
+    Returns a JSON schema `enum` items of a given type.
+    >>> from azul.doctests import assert_json
```
Comment (hannes-ucsc) [inline src/azul/openapi/schema.py]: Empty line after description and each assertion.

File: src/azul/openapi/schema.py [hunk 14]
```diff
@@ -117,6 +122,51 @@ def array(item: TYPE, *items: TYPE, **kwargs):
...
+    {
+        "type": "integer",
+        "format": "int64",
+        "enum": [
+            2,
+            5,
+            7
+        ]
+    }
+    >>> assert_json(enum('bar', 'foo', 0))
```
Comment (hannes-ucsc) [inline src/azul/openapi/schema.py]: assert_json not necessary here

File: src/azul/openapi/schema.py [hunk 15]
```diff
@@ -117,6 +122,51 @@ def array(item: TYPE, *items: TYPE, **kwargs):
...
+    ...
+    AssertionError
+    """
+    if type_ is None:
+        type_ = one(set(map(type, items)))
+    elif isinstance(type_, type):
+        assert all(isinstance(item, type_) for item in items)
+    else:
+        # Can't verify type when string or JSON
+        pass
```
Comment (hannes-ucsc) [inline src/azul/openapi/schema.py]: not covered.

File: src/azul/openapi/params.py [hunk 16]
```diff
@@ -0,0 +1,80 @@
...
+            "format": "int64"
+        }
+    }
+    """
+    return _make_param(name, in_='query', type_=type_, **kwargs)
+
+
+def _make_param(name: str, in_: str, type_: Union[TYPE, schema.optional], **kwargs: PrimitiveJSON) -> JSON:
+    """
+    >>> from azul.doctests import assert_json
```
Comment (hannes-ucsc) [inline src/azul/openapi/params.py]: These doctests add no coverage. Please remove.

File: lambdas/service/app.py [hunk 17]
```diff
@@ -957,7 +959,59 @@ def generate_manifest(event, context):
...
+        params.path('uuid', str, description='UUID of the file to be checked out'),
+        params.query('fileName',
+                     schema.optional(str),
+                     description='The desired name of the file. If absent, the UUID of the file will be used.'),
+        params.query('requestIndex',
+                     schema.optional(int),
+                     description='Number of attempts made through the endpoint to fetch the desired file.'),
+        params.query('wait',
+                     schema.optional(int),
+                     description='If this parameter is 1 and the checkout is still in process, the server will wait '
```
Comment (hannes-ucsc) [inline lambdas/service/app.py]: Wouldn't a triple quoted string be easier here? If indention and newlines are a problem, add a helper to post-process the string.

File: lambdas/service/app.py [hunk 18]
```diff
@@ -957,7 +959,59 @@ def generate_manifest(event, context):
...
 
 
-@app.route('/dss/files/{uuid}', methods=['GET'], cors=True)
+dss_files_spec = {
+    'tags': ['DSS'],
+    'parameters': [
+        params.path('uuid', str, description='UUID of the file to be checked out'),
+        params.query('fileName',
+                     schema.optional(str),
+                     description='The desired name of the file. If absent, the UUID of the file will be used.'),
```
Comment (hannes-ucsc) [inline lambdas/service/app.py]: Don't duplicate code or documentation. Remove the docstring sections after porting them to the spec. Ideally, the entire docstring should be ported and removed alltogether.

File: src/azul/openapi/params.py [hunk 19]
```diff
@@ -0,0 +1,80 @@
...
+from azul.types import (
+    JSON,
+    PrimitiveJSON,
+)
+
+
+# Path params cannot be optional
+def path(name: str, type_: TYPE, **kwargs: PrimitiveJSON) -> JSON:
+    """
+    Returns a JSON `parameters` entry for an endpoint parameter specified via
```
Comment (hannes-ucsc) [inline src/azul/openapi/params.py]: Slightly imprecise and a bit awkward.

```suggestion
    Returns an OpenAPI `parameters` specification of a URL path parameter.
```

See if you can improve the other docstrings accordingly. I haven't look at them in detail.

File: src/azul/openapi/params.py [hunk 20]
```diff
@@ -0,0 +1,80 @@
...
+from azul.openapi.schema import (
+    TYPE
+)
+from azul.types import (
+    JSON,
+    PrimitiveJSON,
+)
+
+
+# Path params cannot be optional
```
Comment (hannes-ucsc) [inline src/azul/openapi/params.py]: I'd move this into the docstring.

File: src/azul/strings.py [hunk 21]
```diff
@@ -46,3 +46,22 @@ def pluralize(word: str, count: int) -> str:
         else:
             result += 's'
     return result
+
+
+def collapse_lines(string: str):
```
Comment (hannes-ucsc) [inline src/azul/strings.py]:
```suggestion
def unwrap(string: str):
```

This will be called a lot so a shorter name makes sense.

File: lambdas/service/app.py [hunk 22]
```diff
@@ -957,57 +962,101 @@ def generate_manifest(event, context):
...
+            'headers': {
+                'Location': responses.header(str, description='This endpoint.'),
+                'Retry-After': responses.header(int, description=collapse_lines(
+                    '''
+                    Recommended number of seconds to wait before requesting the
+                    URL specified in the `Location` header. The response MAY
+                    carry this header even if server-side waiting was requested
+                    via `wait=1`.
+                    '''
+                ))
```
Comment (hannes-ucsc) [inline lambdas/service/app.py]:
```suggestion
                'Retry-After': responses.header(int, description=unwrap('''
                    Recommended number of seconds to wait before requesting the
                    URL specified in the `Location` header. The response MAY
                    carry this header even if server-side waiting was requested
                    via `wait=1`.
                '''))
```

unless `make format` mangles this

Comment (hannes-ucsc) [reply lambdas/service/app.py]: Or even

```suggestion
                'Retry-After': responses.header(int, description=unwrap('''
                    Recommended number of seconds to wait before requesting the
                    URL specified in the `Location` header. The response MAY
                    carry this header even if server-side waiting was requested
                    via `wait=1`.'''))
```

Comment (hannes-ucsc) [reply lambdas/service/app.py]: I'm also wondering if we should build the call to unwrap into the helpers. If a keyword argument named `description` is passed, apply `unwrap()` to it automatically.

Comment (noah-aviel-dove) [reply lambdas/service/app.py]: I think it's a good idea to implement this.
